proposal/first-class-css-functions.md

```diff
@@ -0,0 +1,135 @@
+# First-Class CSS Functions
+
+*([Issue](https://github.com/sass/sass/issues/3176),
+[Changelog](first-class-css-functions.changes.md))*
+
+## Table of Contents
+
+* [Background](#background)
+* [Summary](#summary)
+* [Syntax](#syntax)
+* [Types](#types)
+  * [Serialization](#serialization)
+    * [CssFunction](#cssfunction)
+* [Semantics](#semantics)
+* [Functions](#functions)
+  * [`meta.type-of()`](#metatype-of)
+  * [`meta.fn-name()`](#metafn-name)
+  * [`meta.fn-args()`](#metafn-args)
+
+## Background
+
+> This section is non-normative.
+
+Sass currently treats non-[calculation] [CSS Functions] as fully opaque,
+allowing almost any sequence of tokens within the parentheses and evaluating it
+to an unquoted string. Once a function is used it can't be inspected or
+manipulated in any way.
+
+Notice that interpolation in CSS Functions is already supported and therefore not included in this proposal. `$foo: 1px; var(--foo, $foo);` produces `var(--foo, 1px)`.
+
+[calculation]: https://sass-lang.com/documentation/values/calculations
+[CSS Functions]: https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Functions
+
+## Summary
+
+> This section is non-normative.
+
+This proposal changes all CSS Functions from being parsed as unquoted strings to
+being parsed in-depth and producing a new data type known as a "css-function".
+This data type holds the name and arguments of the function.
+
+For example, given this block of code:
+
+```scss
+$namespace: 'my-pkg'
+
+$foo: var(--foo);
+$bar: var(--bar, 8px);
+$baz: var(--#{$namespace}-baz, 8px);
+$rotation: rotate(0.5turn);
+```
+
+- `meta.type-of($foo)` will return `css-function`
+- `meta.fn-name($foo)` will return `"var"`
+- `meta.fn-args($foo)` will return `unquote("--foo")`
+- `meta.fn-args($bar)` will return `unquote("--bar"), 8px`
+- `meta.fn-args($baz)` will return `unquote("--my-pkg-baz"), 8px`
+- `meta.fn-name($rotation)` will return `"rotate"`
+- `meta.fn-args($rotation)` will return `0.5turn`
+
+Note that if an argument is a number or a nested calculation, it’s returned as
+that type. Otherwise, it’s returned as an unquoted string.
+
+## Syntax
+
+## Types
+
+This proposal introduces a new value type known as a "css-function", with the
+following structure:
+
+```ts
+interface CssFunction {
+  name: string;
+  arguments: CssFunctionArgument[];
+}
+
+type CssFunctionArgument =
+  | Number
+  | UnquotedString
+  | Calculation;
+```
+
+### Serialization
+
+#### CssFunction
+
+To serialize a CSS Function, emit its name followed by "(", then each of its arguments
+separated by ",", then ")".
+
+## Semantics
```

Comment:
Same with the Serialization and Semantics sections, not sure what to put there (feel free to take over instead of just writing comments)